pkg/kv/dist_sender.go

```diff
@@ -1158,51 +1135,54 @@ func (ds *DistSender) sendToReplicas(
-		log.VEventf(ctx, 2, "r%d: sending batch %s to %s",
-			rangeID, args.Summary(), transport.NextReplica())
-	}
+	// Must be unbuffered because tests have blocking SendNext
+	// implementations.
+	done := make(chan BatchCall, 1)
+	log.VEventf(ctx, 2, "r%d: sending batch %s to %s", rangeID, args.Summary(), transport.NextReplica())
 	transport.SendNext(ctx, done)
 
 	// Wait for completions. This loop will retry operations that fail
 	// with errors that reflect per-replica state and may succeed on
 	// other replicas.
-	sendNextTimer := timeutil.NewTimer()
 	slowTimer := timeutil.NewTimer()
-	defer sendNextTimer.Stop()
 	defer slowTimer.Stop()
 	slowTimer.Reset(base.SlowRequestThreshold)
 	for {
-		if timeout, ok := transport.SendNextTimeout(opts.SendNextTimeout); ok {
-			// Only start the send-next timer if we haven't exhausted the transport
-			// (i.e. there is another replica to send to).
-			sendNextTimer.Reset(timeout)
-		}
-
 		select {
-		case <-sendNextTimer.C:
-			sendNextTimer.Read = true
-			// On successive RPC timeouts, send to additional replicas if available.
-			if !transport.IsExhausted() {
-				ds.metrics.SendNextTimeoutCount.Inc(1)
-				log.VEventf(ctx, 2, "timeout, trying next peer: %s", transport.NextReplica())
-				pending++
-				transport.SendNext(ctx, done)
-			}
-
 		case <-slowTimer.C:
 			log.Warningf(ctx, "have been waiting %s sending RPC to r%d for batch: %s",
 				base.SlowRequestThreshold, rangeID, args)
 			ds.metrics.SlowRequestsCount.Inc(1)
 			defer ds.metrics.SlowRequestsCount.Dec(1)
 
 		case call := <-done:
-			pending--
-			err := call.Err
-			if err == nil {
+			if err := call.Err; err != nil {
+				// All connection errors except for an unavailable node (this
+				// is GRPC's fail-fast error), may mean that the request
+				// succeeded on the remote server, but we were unable to
+				// receive the reply. Set the ambiguous commit flag.
+				//
+				// We retry ambiguous commit batches to avoid returning the
+				// unrecoverable AmbiguousResultError. This is safe because
+				// repeating an already-successfully applied batch is
+				// guaranteed to return either a TransactionReplayError (in
+				// case the replay happens at the original leader), or a
+				// TransactionRetryError (in case the replay happens at a new
+				// leader). If the original attempt merely timed out or was
+				// lost, then the batch will succeed and we can be assured the
+				// commit was applied just once.
+				//
+				// The Unavailable code is used by GRPC to indicate that a
+				// request fails fast and is not sent, so we can be sure there
+				// is no ambiguity on these errors. Note that these are common
+				// if a node is down.
+				// See https://github.com/grpc/grpc-go/blob/52f6504dc290bd928a8139ba94e3ab32ed9a6273/call.go#L182
+				// See https://github.com/grpc/grpc-go/blob/52f6504dc290bd928a8139ba94e3ab32ed9a6273/stream.go#L158
+				if haveCommit && grpc.Code(err) != codes.Unavailable {
```

Comment:
This change is not correct. Previously, if we got any RPC error, we continued to try the other replicas, hoping one would succeed. Even if we get an RPC error other than `Unavailable`, if we succeed with the batch at any additional replica, we can still return success instead of the `AmbiguousResultError`. Instead, we're now just fast-failing with the `AmbiguousResultError`.

This is one of the cases that all of this code was trying to solve for. We know that node1 is the leaseholder, we send node1 the batch, it times out because it's behind a network partition, and we retry to a node2 and node3, one of which will be elected Raft leader and gain the lease. Assuming that the original batch didn't succeed at node1, then it will succeed at the new leaseholder (either node2 or node3).